crypto/rand/rand_lib.c

```diff
@@ -831,15 +834,15 @@ void RAND_seed(const void *buf, int num)
 {
     const RAND_METHOD *meth = RAND_get_rand_method();
 
-    if (meth->seed != NULL)
+    if (meth != NULL && meth->seed != NULL)
```

Comment:
I'm feeling a bit uncomfortable about this location here, because there is no way to indicate an error to caller. OTOH: if the CSPRNG is already initialized, the  failure will not be disastrous, and if it is not, the generate call will fail later on.

Reply:
How much of an API break is it to return an int (that current application code would simply ignore)?  If considered harmless, we *could* change the return value for this function (and also for the other functions that are currently void)

Reply:
I didn't think `RAND_get_rand_method` could fail after the RNG was initialised.  Isn't it's failure path when creating the method on first call?

The API break would be when taking the address of the function.  How likely is it for `&RAND_seed` (or just `RAND_seed` since the _&_ is optional) to appear in code?

I think having `RAND_seed` return a failure indicator would be an improvement however.

Reply:
After thinking about it for a while, I don't believe it is worth breaking the `RAND_{add,seed}()` API in order to return an error, for the following reasons:

People currently don't check the return value of `RAND_{add,seed}()`  anyway, as @levitte already noted. And actually it does not matter, because if the (re-)seeding fails then the DRBG will remain in the error state, so `RAND_status()` and `RAND_bytes()` will fail anyway.

The `RAND_{add,seed}()` APIs are obsolescent, because the DRBG reseeds itself anyway. We shouldn't change an API which we don't want to be used anymore.

Reply:
Perhaps some deprecation is in order

Reply:
I figure we'll need to do a big pass over everything at some point and mass deprecate old APIs.

Reply:
I do not think the RAND_add/RAND_seed should be deprecated. They should be clearly declared as unnecessary in documentation but there might be special environments and applications where such calls can be handy. Just looking at rand_unix.c what about the case of OPENSSL_SYS_VXWORKS or OPENSSL_SYS_UEFI?

Reply:
I'm sceptical that applications will pass decent randomness to `RAND_{add,seed}`...  I've seen far too many poor attempts that were about as good as doing nothing.

With a HW RNG, it's possible but if the OS doesn't supply any/enough, where will a user find more?

Reply:
Having the call available means that if you have one application for a special device where openssl is compilable but does not have implemented any method to poll for entropy in the current sources, but there is for example HW RNG source, you can still workaround this just by the call from application. Otherwise it would require patching the openssl code. What we would get from the deprecation when we will have to have some call like this internally for the RAND_poll anyway.